File: python/smoke_verify/verify.py
"""Chain verification. Fail-closed.

A chain is valid only if EVERY entry passes EVERY check; the first failure
stops verification and is reported with its index + reason. Checks per entry i:

  1. sequence == i                                  (monotonic, no gaps/reorder)
  2. prev_entry_hash == prev recomputed entry_hash  (chain link; null at genesis)
  3. recomputed entry_hash == stored entry_hash     (record not mutated)
  4. ECDSA-P256 verify over signed_digest(entry_hash) under the header key
  5. envelope.public_key + record.signer_key_id bind to the header key
"""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Optional, Union

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import (
    Prehashed,
    encode_dss_signature,
)

from .envelope import AttestationEnvelopeV0, ChainHeaderV0
from .schema import AttestationError
from .keys import spki_fingerprint

__all__ = ["VerifyResult", "verify_chain", "verify_log"]


@dataclass(frozen=True)
class VerifyResult:
    ok: bool
    count: int                      # entries checked (excludes header)
    broken_index: Optional[int]     # 0-based entry index of first failure, else None
    reason: Optional[str]
    key_id: Optional[str] = None
    ended: bool = False             # last entry is a session_end record

    def __bool__(self) -> bool:
        return self.ok


def _fail(count: int, idx: Optional[int], reason: str, key_id: Optional[str] = None) -> VerifyResult:
    return VerifyResult(ok=False, count=count, broken_index=idx, reason=reason, key_id=key_id)


def verify_chain(
    header: ChainHeaderV0,
    envelopes: list[AttestationEnvelopeV0],
    *,
    trusted_spki_sha256: Optional[str] = None,
    trusted_key_id: Optional[str] = None,
) -> VerifyResult:
    """Verify a chain. Internal consistency only proves the log is consistent
    under WHATEVER key it carries — pass `trusted_spki_sha256` (and/or
    `trusted_key_id`) to also prove it was signed by the EXPECTED key. Without a
    pin, an attacker who rewrites the log and signs it with their own key
    produces a self-consistent but untrusted chain that internal checks accept.
    """
    # Trust-anchor pinning (external authenticity), checked before anything else.
    if trusted_spki_sha256 is not None and header.spki_sha256 != trusted_spki_sha256:
        return _fail(0, None,
                     f"header key {header.spki_sha256[:16]}… is not the trusted anchor "
                     f"{trusted_spki_sha256[:16]}…", header.key_id)
    if trusted_key_id is not None and header.key_id != trusted_key_id:
        return _fail(0, None,
                     f"header key_id {header.key_id!r} != trusted key_id {trusted_key_id!r}",
                     header.key_id)

    # Load the chain's public key from the header and sanity-check its fingerprint.
    try:
        spki_der = bytes.fromhex(header.spki_der_hex)
        pub = serialization.load_der_public_key(spki_der)
    except Exception as e:  # noqa: BLE001 - malformed header key is a hard fail
        return _fail(0, None, f"header public key unreadable: {e}", header.key_id)
    if not isinstance(pub, ec.EllipticCurvePublicKey):
        return _fail(0, None, "header key is not an EC public key", header.key_id)
    if spki_fingerprint(spki_der) != header.spki_sha256:
        return _fail(0, None, "header spki_sha256 does not match spki_der", header.key_id)

    prev_eh: Optional[str] = None
    for i, env in enumerate(envelopes):
        rec = env.record

        if rec.sequence != i:
            return _fail(i, i, f"sequence {rec.sequence} != expected {i}", header.key_id)
        if rec.prev_entry_hash != prev_eh:
            return _fail(i, i, "prev_entry_hash does not link to previous entry", header.key_id)

        recomputed = rec.entry_hash_hex()
        if recomputed != env.entry_hash:
            return _fail(i, i, "stored entry_hash != recomputed (record mutated)", header.key_id)

        if env.public_key.spki_sha256 != header.spki_sha256:
            return _fail(i, i, "envelope public_key fingerprint != header key", header.key_id)
        if env.public_key.key_id != header.key_id or rec.signer_key_id != header.key_id:
            return _fail(i, i, "key_id mismatch (envelope/record vs header)", header.key_id)
        # The header is unsigned metadata; binding each signed record's session_id
        # to it makes a header swap an explicit failure rather than silent.
        if rec.session_id != header.session_id:
            return _fail(i, i, "record session_id != chain header session_id", header.key_id)

        try:
            der_sig = encode_dss_signature(int(env.signature.r, 16), int(env.signature.s, 16))
            pub.verify(der_sig, rec.signed_digest(), ec.ECDSA(Prehashed(hashes.SHA256())))
        except InvalidSignature:
            return _fail(i, i, "signature does not verify", header.key_id)
        except Exception as e:  # noqa: BLE001 - malformed r/s etc.
            return _fail(i, i, f"signature check error: {e}", header.key_id)

        prev_eh = recomputed

    ended = bool(envelopes) and envelopes[-1].record.event_type == "session_end"
    return VerifyResult(ok=True, count=len(envelopes), broken_index=None,
                        reason=None, key_id=header.key_id, ended=ended)


def verify_log(
    path: Union[str, Path],
    *,
    trusted_spki_sha256: Optional[str] = None,
    trusted_key_id: Optional[str] = None,
) -> VerifyResult:
    """Verify a `.jsonl` attestation log: line 1 = header, rest = envelopes.

    A malformed/truncated line (bad JSON, partial last line) fails closed at
    that entry index (STRICT: a truncated final line is a verification
    failure, never silently ignored).

    Pass `trusted_spki_sha256` (and/or `trusted_key_id`) to require the log was
    signed by an externally-pinned key, not merely self-consistent under the
    key it carries. Callers making an authenticity claim SHOULD pin."""
    p = Path(path)
    try:
        raw = p.read_text(encoding="utf-8")
    except OSError as e:
        return _fail(0, None, f"cannot read log: {e}")

    lines = [ln for ln in raw.splitlines() if ln.strip()]
    if not lines:
        return _fail(0, None, "empty log (no header)")

    try:
        header = ChainHeaderV0.from_json_line(lines[0])
    except (AttestationError, ValueError) as e:
        return _fail(0, None, f"bad chain header: {e}")

    envelopes: list[AttestationEnvelopeV0] = []
    for j, line in enumerate(lines[1:]):
        try:
            envelopes.append(AttestationEnvelopeV0.from_json_line(line))
        except (AttestationError, ValueError) as e:
            # j is the entry index (0-based) of the offending envelope line.
            return _fail(len(envelopes), j, f"malformed/truncated entry: {e}", header.key_id)

    return verify_chain(
        header, envelopes,
        trusted_spki_sha256=trusted_spki_sha256,
        trusted_key_id=trusted_key_id,
    )

File: ts/src/verify.ts
// smoke.attestation.v0 — chain verifier (TypeScript port). Fail-closed.
//
// Mirrors the Python verifier. Zero deps: ECDSA-P256 verification uses Node's
// built-in crypto. Because signed_digest = SHA256(DOMAIN || 0x00 || entry_hash),
// we verify by letting crypto hash the known message M = DOMAIN||0x00||entry_hash
// with SHA-256 (no prehashed-verify needed) against the raw r||s signature.
import { createPublicKey, verify as cryptoVerify } from "node:crypto";
import { readFileSync } from "node:fs";

import { DOMAIN_TAG, entryHash, parseLossless, sha256 } from "./canonical.ts";

export interface VerifyResult {
  ok: boolean;
  count: number;
  brokenIndex: number | null;
  reason: string | null;
  keyId: string | null;
  ended: boolean;
}

export interface VerifyOpts {
  /** The trust anchor: expected signing key fingerprint (SHA-256 of SPKI DER). */
  trustedSpkiSha256?: string;
  /**
   * SECURITY: a FILTER on the header's key_id LABEL, not a trust anchor —
   * a forger picks their own key_id, so matching it proves nothing about the
   * key. Use only as an additional constraint beside trustedSpkiSha256.
   */
  trustedKeyId?: string;
}

function fail(count: number, idx: number | null, reason: string, keyId: string | null = null): VerifyResult {
  return { ok: false, count, brokenIndex: idx, reason, keyId, ended: false };
}

// The 14 record fields (must be present, no more, no fewer) — mirrors the Python
// verifier's strict from_dict so both verifiers reject the SAME inputs.
const RECORD_FIELDS = [
  "version", "event_id", "session_id", "sequence", "event_type", "tool_name",
  "tool_input_sha256", "tool_output_sha256", "timestamp_unix_ns", "actor", "cwd",
  "repo", "prev_entry_hash", "signer_key_id",
];

/** Structural validation before any crypto: returns an error string or null. */
function envelopeShapeError(env: any): string | null {
  if (env == null || typeof env !== "object") return "envelope is not an object";
  for (const k of ["record", "entry_hash", "signature", "public_key"]) {
    if (!(k in env)) return `missing ${k}`;
  }
  const rec = env.record;
  if (rec == null || typeof rec !== "object") return "record is not an object";
  for (const f of RECORD_FIELDS) if (!(f in rec)) return `record missing ${f}`;
  for (const f of Object.keys(rec)) if (!RECORD_FIELDS.includes(f)) return `record has unknown field ${f}`;
  for (const k of ["alg", "r", "s", "encoding"]) if (!(k in env.signature)) return `signature missing ${k}`;
  for (const k of ["key_id", "alg", "spki_sha256"]) if (!(k in env.public_key)) return `public_key missing ${k}`;
  return null;
}

export function verifyChain(header: any, envelopes: any[], opts: VerifyOpts = {}): VerifyResult {
  if (header == null || header.type !== "smoke-attest-header") return fail(0, null, "not a chain header");
  const keyId: string = header.key_id;

  if (opts.trustedSpkiSha256 != null && header.spki_sha256 !== opts.trustedSpkiSha256) {
    return fail(0, null, "header key is not the trusted anchor (spki_sha256 mismatch)", keyId);
  }
  if (opts.trustedKeyId != null && header.key_id !== opts.trustedKeyId) {
    return fail(0, null, `header key_id ${header.key_id} != trusted key_id ${opts.trustedKeyId}`, keyId);
  }

  let spkiDer: Buffer;
  let pub: ReturnType<typeof createPublicKey>;
  try {
    spkiDer = Buffer.from(header.spki_der_hex, "hex");
    pub = createPublicKey({ key: spkiDer, format: "der", type: "spki" });
  } catch (e) {
    return fail(0, null, `header public key unreadable: ${e}`, keyId);
  }
  if (sha256(spkiDer).toString("hex") !== header.spki_sha256) {
    return fail(0, null, "header spki_sha256 does not match spki_der", keyId);
  }

  // A header-only log carries zero signed statements (headers are unsigned),
  // and the writer always emits session_start with the header — a legitimate
  // 0-entry log does not exist. Fail closed.
  if (envelopes.length === 0) {
    return fail(0, null,
      "chain has a header but no entries — an empty log proves nothing and is never a legitimate state (fail-closed)",
      keyId);
  }

  let prev: string | null = null;
  for (let i = 0; i < envelopes.length; i++) {
    const env = envelopes[i];
    const shapeErr = envelopeShapeError(env);
    if (shapeErr) return fail(i, i, `malformed/truncated entry: ${shapeErr}`, keyId);
    const rec = env.record;

    if (rec.sequence !== BigInt(i)) return fail(i, i, `sequence ${rec.sequence} != expected ${i}`, keyId);
    const expectedPrev = i === 0 ? null : prev;
    if ((rec.prev_entry_hash ?? null) !== expectedPrev) {
      return fail(i, i, "prev_entry_hash does not link to previous entry", keyId);
    }

    const ehHex = entryHash(rec).toString("hex");
    if (ehHex !== env.entry_hash) return fail(i, i, "stored entry_hash != recomputed (record mutated)", keyId);

    if (env.public_key.spki_sha256 !== header.spki_sha256) {
      return fail(i, i, "envelope public_key fingerprint != header key", keyId);
    }
    if (env.public_key.key_id !== keyId || rec.signer_key_id !== keyId) {
      return fail(i, i, "key_id mismatch (envelope/record vs header)", keyId);
    }
    // Bind each signed record to the unsigned header's session_id.
    if (rec.session_id !== header.session_id) {
      return fail(i, i, "record session_id != chain header session_id", keyId);
    }

    const eh = Buffer.from(ehHex, "hex");
    const msg = Buffer.concat([DOMAIN_TAG, Buffer.from([0]), eh]); // signed_digest = SHA256(msg)
    let sig: Buffer;
    try {
      sig = Buffer.concat([Buffer.from(env.signature.r, "hex"), Buffer.from(env.signature.s, "hex")]);
    } catch (e) {
      return fail(i, i, `bad signature encoding: ${e}`, keyId);
    }
    const sigOk = cryptoVerify("sha256", msg, { key: pub, dsaEncoding: "ieee-p1363" }, sig);
    if (!sigOk) return fail(i, i, "signature does not verify", keyId);

    prev = ehHex;
  }

  const ended = envelopes.length > 0 && envelopes[envelopes.length - 1].record.event_type === "session_end";
  return { ok: true, count: envelopes.length, brokenIndex: null, reason: null, keyId, ended };
}

export function verifyLog(path: string, opts: VerifyOpts = {}): VerifyResult {
  let text: string;
  try {
    text = readFileSync(path, "utf8");
  } catch (e) {
    return fail(0, null, `cannot read log: ${e}`);
  }
  const lines = text.split("\n").filter((l) => l.trim() !== "");
  if (lines.length === 0) return fail(0, null, "empty log (no header)");

  let header: any;
  try {
    header = parseLossless(lines[0]);
  } catch (e) {
    return fail(0, null, `bad chain header: ${e}`);
  }
  const envelopes: any[] = [];
  for (let j = 1; j < lines.length; j++) {
    try {
      envelopes.push(parseLossless(lines[j]));
    } catch (e) {
      return fail(envelopes.length, j - 1, `malformed/truncated entry: ${e}`, header?.key_id ?? null);
    }
  }
  return verifyChain(header, envelopes, opts);
}
